src/edu/washington/escience/myria/perfenforce/PerfEnforceOnlineLearning.java

```diff
@@ -0,0 +1,396 @@
+   */
+  public void trainOnlineQueries(final int configurationIndex) throws PerfEnforceException {
+    String MOAFileName = Paths.get(onlineLearningPath, "moa.jar").toString();
+    String trainingFileName = Paths.get(onlineLearningPath, "training.arff").toString();
+    String modifiedTrainingFileName =
+        Paths.get(onlineLearningPath, "training-modified-" + configurationIndex + ".arff")
+            .toString();
+    String predictionsFileName =
+        Paths.get(onlineLearningPath, "predictions" + configurationIndex + ".txt").toString();
+
+    try {
+      PrintWriter outputWriter = new PrintWriter(modifiedTrainingFileName);
+      outputWriter.close();
+
+      // copy training file to new file
+      FileChannel src = new FileInputStream(trainingFileName).getChannel();
+      FileChannel dest = new FileOutputStream(modifiedTrainingFileName).getChannel();
+      dest.transferFrom(src, 0, src.size());
+      src.close();
+      dest.close();
+
+      // Append all previous data points
+      FileWriter appendDataWriter = new FileWriter(modifiedTrainingFileName, true);
+      for (String s : previousDataPoints) {
+        appendDataWriter.write(s + "\n");
+      }
+
+      // Append the current point
+      String newPoint = getQueryFeature(configurationIndex, 0);
+      appendDataWriter.write(newPoint + "\n");
+      appendDataWriter.close();
+
+      String moaCommand =
+          String.format(
+              "EvaluatePrequentialRegression -l (rules.functions.Perceptron  -d -l %s) -s (ArffFileStream -f %s) -e (WindowRegressionPerformanceEvaluator -w 1) -f 1 -o %s",
+              .04,
+              modifiedTrainingFileName,
+              predictionsFileName);
+
+      String[] arrayCommand =
+          new String[] {"java", "-classpath", MOAFileName, "moa.DoTask", moaCommand};
+
+      Process p = Runtime.getRuntime().exec(arrayCommand);
+
+      BufferedReader reader = new BufferedReader(new InputStreamReader(p.getInputStream()));
+      while ((reader.readLine()) != null) {}
+
+      parsingOnlineFile(configurationIndex, predictionsFileName);
+    } catch (Exception e) {
+      throw new PerfEnforceException("Error during training");
+    }
+  }
+
+  /**
+   * Helps parse the output file from MOA to read the runtime prediction
+   *
+   * @param configurationIndex the configuration size in consideration
+   * @param predictionFileName the file that contains the prediction
+   */
+  public void parsingOnlineFile(final int configurationIndex, final String predictionFileName)
+      throws PerfEnforceException {
+    try {
+      BufferedReader streamReader = new BufferedReader(new FileReader(predictionFileName));
+      String currentLine = "";
+      double nextQueryPrediction = 0;
+      while ((currentLine = streamReader.readLine()) != null) {
+        nextQueryPrediction = Double.parseDouble((currentLine.split(",")[0]).split(":")[1]);
+      }
+      streamReader.close();
```

Comment:
Put it in `finally` is safer. (Also there are several places with the similar issue.)